libsplinter/src/node_registry/yaml/mod.rs

```diff
@@ -117,11 +116,14 @@ impl NodeRegistryReader for YamlNodeRegistry {
 }
 
 impl NodeRegistryWriter for YamlNodeRegistry {
-    fn add_node(&self, node: Node) -> Result<(), NodeRegistryError> {
+    fn add_or_update_node(&self, node: Node) -> Result<(), NodeRegistryError> {
         let mut nodes = self
             .get_cached_nodes()
             .map_err(|err| NodeRegistryError::InternalError(Box::new(err)))?;
 
+        // If a node with the same identity already exists, remove it
+        nodes.retain(|existing_node| existing_node.identity != node.identity);
```

Comment:
Shouldn't this be done after the validation?

Reply:
Good catch; fixed.

Reply:
Actually, no; when checking for uniqueness of the node's endpoint, we don't want to be comparing the node against itself. Removing the node from this vec has no consequences if the method returns early, because it's just a cache that only gets written at the end of the method.

Reply:
You can at least validate that the incoming node is valid before removing the old node from the cache.

Reply:
Moved the field validation before removing the node from the cache.